### PR TITLE
feat(operate): new exporter module

### LINKED ISSUE #18304: Port importer logic to exporter
### Description

To start with converting importer to exporter, we need to copy the current importer logic to new Zeebe exporter. It must not be ideal and take into account all the changes that we plan to perform within this epic, but rather do exactly the same as importer is doing now.

### TODO

- [x] create new Zeebe exporter module: PR #18886
- [x] implement simple code structure to transform Zeebe records to Operate entities and persist them in Operate indices: the same PR #18886
- [x] implement BatchRequest using the new ES client https://github.com/camunda/camunda/issues/19108

Copy all business logic from `*ZeebeRecordProcessor` classes to use in record transformation: 
- [x] #18908
- [x] #18911
- [x] #18912
- [x] #18913
- [x] #18914
- [x] #18915
- [x] #19387
- [x] #22642

### Additional Information

* performance is not a priority in this issue
* [prototype code](https://github.com/camunda/operate/issues/6031) can be used when needed, but importer logic must be copied from current `main`

---------------------------------------------------------------------------------------------

### Acceptance Criteria

- [ ] Checked if the specific data needs to be provided for the QA process

### Definition of Ready - Checklist

- [ ] The issue has a meaningful title, description and testable acceptance criteria
- [ ] If documentation needs to be updated, an issue is created in the [camunda-docs](https://github.com/camunda/camunda-docs) repo, and the issue is added to our Operate project board.
- [ ] If HELM charts need to be updated, an issue is created in the [camunda-platform-heml](https://github.com/camunda/camunda-platform-helm) repo, and the issue is added to our Operate project board.
- [ ] Potential Security risks have been considered
- [ ] Cross-team dependencies have been considered

### PR DESCRIPTION
**Pull request by sdorokhova**
## Description

* new exporter logic to write data directly to Operate indices
* very simply structure to start from
* one sample data handler that converts and flushes process definitions `ProcessHandler`

## Related issues

related to #18304

### CODE REVIEW

**Comment by sdorokhova** [comment]
To try the thing one can:
* run Elasticsearch in docker
* run `io.camunda.application.StandaloneBroker` class with following configuration parameteres:
```
ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_ARGS_BULK_SIZE=1;ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_ARGS_INDEX_DEPLOYMENT=false;ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_CLASSNAME=io.camunda.zeebe.exporter.ElasticsearchExporter;ZEEBE_BROKER_EXPORTERS_OPERATE_ARGS_BULK_SIZE=1;ZEEBE_BROKER_EXPORTERS_OPERATE_ARGS_ELASTICSEARCH_INDEXPREFIX=newoperate;ZEEBE_BROKER_EXPORTERS_OPERATE_CLASSNAME=io.camunda.zeebe.operate.exporter.OperateElasticsearchExporter
```
* deploy process from the modeler
* check Elastic indices

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ElasticsearchConnector.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ElasticsearchConnector.java:109]: If we don't need this commented-out code I think we should remove it

File: operate/schema/src/main/java/io/camunda/operate/schema/indices/AbstractIndexDescriptor.java
Comment (matt-whiteman) [inline operate/schema/src/main/java/io/camunda/operate/schema/indices/AbstractIndexDescriptor.java:51]: Is it possible for more than one thread to call either of these methods on the same index descriptor? If so, we need to take that into account to avoid the edge case where one thread calls set/get while another one is calling set.

If we do have a multi-threaded possibility, recommend:
- Make the get/set method synchronized
- Add volatile keyword to indexPrefix

File: operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/NewElasticsearchBatchRequest.java
Comment (matt-whiteman) [inline operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/NewElasticsearchBatchRequest.java:26]: Just leaving a note that we should consider a better name in the future. I assume we are going to remove the old ElasticsearchBatchRequest at some point?

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ElasticsearchConnector.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ElasticsearchConnector.java:48]: Understood that this is a copy/paste for now. When we are ready to refactor I have some suggestions and would like to have a discussion

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java:29]: Can you expand or clarify this 'static' comment (and the 'dynamic' one below)? It's unclear exactly what it's referring to and 'static' can be confusing since that's a java keyword and the variable is not static.

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java:43]: This is a bit difficult to follow and understand exactly what this stream is doing, a comment explaining what exactly is happening in this stream would be helpful

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java:84]: Suggest a null check on both left and right, I believe we had an edge case of a NPE involving tuples in the exporter recently

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java:93]: is there a situation where multiple threads might call this? Or one thread calls reset() while another thread is in another method utilizing the cache?

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/OperateElasticsearchExporter.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/OperateElasticsearchExporter.java:28]: Suggestion: This class is tightly coupled to its various internal components (client, writer, controller, etc). Instead of private create methods within this class, create separate factory/builder classes that are passed into the constructor and delegate the logic there. This allows for better testing of this particular class in isolation.

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/handlers/ProcessHandler.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/handlers/ProcessHandler.java:41]: Can we possibly make this a parameter to the constructor or function that uses it? XMLUtil looks fairly complex and creating it directly in here doesn't allow us to mock in tests

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/util/XMLUtil.java
Comment (matt-whiteman) [inline zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/util/XMLUtil.java:28]: Do we plan to make this a spring bean later on? This feels like a shared component and passing a single instance to the various handlers rather than having each handler create and manage its own instance would make things easier from a management and testing standpoint

**Comment by matt-whiteman** [comment]
Some of the new classes look unit testable in their current state, are there plans to add any before starting on the subtasks? Would be good to have these classes that are templates and going to play a major role in other parts of the exporter solid and tested (and any changes made early on)

File: operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/NewElasticsearchBatchRequest.java
Comment (sdorokhova) [reply operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/NewElasticsearchBatchRequest.java:26]: I really could not find the better name. Deprecated client is called `RestHighLevelClient` and Operate class that uses it is called `ElasticsearchBatchRequest`.

New client is called `ElactisearchClient` and I was not sure how to call the corresponding BatchRequest
implementation. 

Should we may be rename `ElasticsearchBatchRequest -> RestHighLevelElasticsearchBatchRequest` and name the new one `ElasticsearchBatchRequest` ?

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ElasticsearchConnector.java
Comment (sdorokhova) [reply zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ElasticsearchConnector.java:109]: We will need it when converting to Spring, but I will remove for now 👍

File: operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/NewElasticsearchBatchRequest.java
Comment (matt-whiteman) [reply operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/NewElasticsearchBatchRequest.java:26]: I think that would be a good idea, it makes it clear too that it's intended for (or uses) the RestHighLevelClient. We could even mark it deprecated as well if the plan is to eventually remove it or get away from it completely

File: zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ElasticsearchConnector.java
Comment (matt-whiteman) [reply zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ElasticsearchConnector.java:109]: If we want to keep the function for now to re-use when spring is added that's fine, let's just add a comment that explains that

**Comment by matt-whiteman** [comment]
One of the ITs is failing, once we can fix that I can approve if you are ready

File: operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/NewElasticsearchBatchRequest.java
Comment (mihail-ca) [reply operate/schema/src/main/java/io/camunda/operate/store/elasticsearch/NewElasticsearchBatchRequest.java:26]: Full implementation here: https://github.com/camunda/camunda/pull/19111

**Comment by matt-whiteman** [comment]
Since several of these suggestions may be covered when we refactor to add Spring in, feel free to just make a note and we don't have to do them for this PR (assuming we want to do them at all)

**Comment by matt-whiteman** [comment]
Changes I made:
- Adjusted how indexPrefix is stored so the existing integration test passes without changing needed behavior for the new exporter
- Fixed a linting issue with one of the commit messages
- Rebased from feature branch to merge in https://github.com/camunda/camunda/pull/19111, made adjustments to compensate for the new constructor signature
- Updated how XMLUtil and ElasticsearchScriptBuilder are created and passed around to aid in creating unit tests later

**Review by ghost — APPROVED** [review]
@matt-whiteman  I think we can merge this as it is, to have it in the feature branch as starting point. We can make adjustments afterwards.

**Comment by matt-whiteman** [comment]
> @matt-whiteman I think we can merge this as it is, to have it in the feature branch as starting point. We can make adjustments afterwards.

@mihail-ca Agree, I'll get it merged in now

**Comment by sdorokhova** [comment]
@mihail-ca @matt-whiteman thank you for taking care of this! I will go through all the proposal and create the separate issue(s) to take them into account

**Pull request merged**
